### PR TITLE
move from python-lzo to lzallright

### PR DESCRIPTION
**Pull request by qkaiser**
In order to run unblob everywhere, we decided to move away from python-lzo because:
- it doesn't have a wheel, so you still need a C compiler and lzo sources
- licensing issues

This branch replace python-lzo with https://github.com/henriquegemignani/py-lzokay, a Python binding for https://crates.io/crates/lzokay

### CODE REVIEW

**Review by vlaci — APPROVED** [review]
I am okay with this, however I am strugling to get py-lzokay compiled for nix. There is some cython black magic we need to sort out. 

As jefferson is included in nixos, it would be good citizenship to figure out how to build it before releasing a new version.

**Comment by qkaiser** [comment]
Agreed. I can keep it in draft until you have more info about Nix.

**Comment by vlaci** [comment]
> Agreed. I can keep it in draft until you have more info about Nix.

I tried to buoild it in a naive way following the steps in the github action, but failed yet again.

<details>
<summary>Build result...</summary>
```console
/tmp/tmp.Qcbe6lejqA                                                                                                                                                                                            
❯ git clone https://github.com/henriquegemignani/py-lzokay/ --recursive
Cloning into 'py-lzokay'...
remote: Enumerating objects: 143, done.
remote: Counting objects: 100% (143/143), done.
remote: Compressing objects: 100% (69/69), done.
remote: Total 143 (delta 59), reused 115 (delta 35), pack-reused 0
Receiving objects: 100% (143/143), 27.22 KiB | 3.89 MiB/s, done.
Resolving deltas: 100% (59/59), done.
Submodule 'native/lzokay' (https://github.com/jackoalan/lzokay) registered for path 'native/lzokay'
Cloning into '/tmp/tmp.Qcbe6lejqA/py-lzokay/native/lzokay'...
remote: Enumerating objects: 53, done.        
remote: Total 53 (delta 0), reused 0 (delta 0), pack-reused 53        
Receiving objects: 100% (53/53), 16.89 KiB | 1.05 MiB/s, done.
Resolving deltas: 100% (26/26), done.
Submodule path 'native/lzokay': checked out '546a9695271e8a8b4711383f828172754fd825f2'
❯ cd py-lzokay

/tmp/tmp.Qcbe6lejqA/py-lzokay  main                                                                                                                                                                          
❯ python -m venv .venv                                                        
❯ .venv/bin/pip install build
Collecting build
  Downloading build-0.10.0-py3-none-any.whl (17 kB)
Collecting tomli>=1.1.0
  Using cached tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting packaging>=19.0
  Using cached packaging-23.1-py3-none-any.whl (48 kB)
Collecting pyproject_hooks
  Using cached pyproject_hooks-1.0.0-py3-none-any.whl (9.3 kB)
Installing collected packages: tomli, packaging, pyproject_hooks, build
Successfully installed build-0.10.0 packaging-23.1 pyproject_hooks-1.0.0 tomli-2.0.1

[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python -m pip install --upgrade pip
❯ .venv/bin/pip install -U pip
Requirement already satisfied: pip in ./.venv/lib/python3.10/site-packages (23.0.1)
Collecting pip
  Using cached pip-23.1.2-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.0.1
    Uninstalling pip-23.0.1:
      Successfully uninstalled pip-23.0.1
Successfully installed pip-23.1.2
❯ .venv/bin/python -m build --wheel
* Creating venv isolated environment...
* Installing packages in isolated environment... (Cython>=3.0a10, setuptools>=44.0.0, setuptools_scm[toml]>=3.4, wheel>=0.34.2)
* Getting build dependencies for wheel...
Compiling /tmp/tmp.Qcbe6lejqA/py-lzokay/_lzokay.pyx because it changed.
[1/1] Cythonizing /tmp/tmp.Qcbe6lejqA/py-lzokay/_lzokay.pyx
running egg_info
creating lzokay.egg-info
writing lzokay.egg-info/PKG-INFO
writing dependency_links to lzokay.egg-info/dependency_links.txt
writing top-level names to lzokay.egg-info/top_level.txt
writing manifest file 'lzokay.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*' under directory 'Source'
adding license file 'LICENSE'
writing manifest file 'lzokay.egg-info/SOURCES.txt'
* Installing packages in isolated environment... (wheel)
* Building wheel...
running bdist_wheel
running build
running build_py
creating build
creating build/lib.linux-x86_64-cpython-310
creating build/lib.linux-x86_64-cpython-310/lzokay
copying lzokay/version.py -> build/lib.linux-x86_64-cpython-310/lzokay
copying lzokay/__init__.py -> build/lib.linux-x86_64-cpython-310/lzokay
running egg_info
writing lzokay.egg-info/PKG-INFO
writing dependency_links to lzokay.egg-info/dependency_links.txt
writing top-level names to lzokay.egg-info/top_level.txt
reading manifest template 'MANIFEST.in'
warning: no files found matching '*' under directory 'Source'
adding license file 'LICENSE'
writing manifest file 'lzokay.egg-info/SOURCES.txt'
error: Error: setup script specifies an absolute path:

    /tmp/tmp.Qcbe6lejqA/py-lzokay/_lzokay.cpp

setup() arguments must *always* be /-separated paths relative to the
setup.py directory, *never* absolute paths.


ERROR Backend subprocess exited when trying to invoke build_wheel
.venv/bin/python -m build --wheel  6,22s user 0,50s system 96% cpu 7,005 total
```
</details>

**Comment by vlaci** [comment]
An alternative https://pypi.org/project/lzallright/ :upside_down_face:

**Comment by qkaiser** [comment]
> An alternative https://pypi.org/project/lzallright/ 🙃

ahahah, I switched to it. Works like a charm.

File: jefferson/jffs2.py [hunk 1]
```diff
@@ -180,7 +180,8 @@ def unpack(self, data):
             elif self.compr == JFFS2_COMPR_LZMA:
                 self.data = jffs2_lzma.decompress(node_data, self.dsize)
             elif self.compr == JFFS2_COMPR_LZO:
-                self.data = lzo.decompress(node_data, False, self.dsize)
+                compressor = LZOCompressor()
+                self.data = compressor.decompress(node_data)
```
Comment (vlaci) [inline jefferson/jffs2.py:184]: It is a static method, no need to instantiate:
```suggestion
                self.data = LZOCompressor.decompress(node_data)
```

Comment (qkaiser) [reply jefferson/jffs2.py]: Fixed

**Comment by stkw0** [comment]
Seems that lzallright depends on lzokay which is unmaintained. Is there some other alternative that can be used? 
For now to package jefferson for Gentoo I  needed to revert this PR.
I would prefer to open an issue but I didn't see any place to report issuesl

Thank you very much.

**Comment by qkaiser** [comment]
> Seems that lzallright depends on lzokay which is unmaintained.

What do you mean it's unmaintained ? I don't see outstanding issues in their tracker or stale pull requests.

> For now to package jefferson for Gentoo I needed to revert this PR.

Why ?

**Comment by stkw0** [comment]
If it's not strictly unmaintained, it's at least immature. The last commit of lzokay-rs was 3 years ago which is a wrapper arround lzokay which don't have any tags nor releases (which is very useful from packaging perspective). and it has very few commits, the last three commits where 5 months ago, 7 months and 4 years ago. 
This is a hint that probably, if packaged, will need to apply downstream patches for both packages. 

I don't proper understand the license issues here. However, from a packaging perspective, the chain of dependencies is problematic. python-lzo instead seems to be well maintained and doesn't depend on problematic packages.

**Comment by vlaci** [comment]
We are not using `lzokay-rs`. [`lzallright`](https://github.com/vlaci/lzallright) has the latest commits from [`lzokay`](https://github.com/AxioDL/lzokay) directly, and does not depend on it at all.

**Comment by stkw0** [comment]
That's good to not depen on lzokay-rs, I seen it mentioned in the PR and didn't properly read the lzallright code. There is still an issue in that lzallright seems to bundle lzokay and in most distros there is a policy of not bundling libraries. Thus, requiring to create a different package for lzokay if possible.

**Comment by vlaci** [comment]
> That's good to not depen on lzokay-rs, I seen it mentioned in the PR and didn't properly read the lzallright code. There is still an issue in that lzallright seems to bundle lzokay and in most distros there is a policy of not bundling libraries. Thus, requiring to create a different package for lzokay if possible.

I understand, that this is a hurdle for distro maintainers, but as python packaging being the mess it is, I am in favor of keeping everything statically linked for easier downstream usage. 

My 2c would be in favor of treating lzallright as its standalone package and it reusing lzokay code is an implementation detail If this would be a blocker for Gentoo we can come up with something.

**Comment by stkw0** [comment]
I see, it would be very nice to have an actively maintained rust library for that. About the licensing issue, I guess it would be an issue if python-lzo is embeded into jefferson or unblob. If it's used as a separated library/program which jefferson depends on and uses, then to my understanding it will be all fine. I suppose the issue arrise because the product you are developing requires it to also bundle and distribute python-lzo with it, which is not possible due to the license, but that doesn't apply to the distro level. If that's the case, would it be possible to have an env flag or similar to decide between using python-lzo or lzallright?

**Comment by stkw0** [comment]
> I understand, that this is a hurdle for distro maintainers, but as python packaging being the mess it is, I am in favor of keeping everything statically linked for easier downstream usage.
> 
> My 2c would be in favor of treating lzallright as its standalone package and it reusing lzokay code is an implementation detail If this would be a blocker for Gentoo we can come up with something.

I understand the decision of statically link everything. But on Gentoo there have been a lot of effort to fix as much as possible the situation which python packaging. So I would prefer to have the option to use python. At the same time though, there can be an option to use jefferson with lzallright so the user can decide according to his needs (or switch between a potential new experimental decompressor in rust and the old one).

**Pull request merged**